### PR TITLE
Setting up basic authentication

### PR DESCRIPTION
**Pull request by styliii**
Using Devise with JWT's.

TODO
1. set up master in production so it can decrypt the credentials file
2. scope workflow endpoints such that current user only has access to resources associated to user
3. update mailer for resetting password email

### CODE REVIEW

File: config/credentials.yml.enc [hunk 1]
```diff
@@ -0,0 +1 @@
+zhSha3wdGy16LXlNk2OrE7iOJi6oUv397vWzVPDJyi5sYcAiL6rYcgIYRKentmndWDlUSdDLavJ9EnI3iyQXTkH99Ll5k/3v7uVclxJTapU6vZr1HwUy0bECw+O7HfnMJ/0APcgmh0dedaOyKFg50u3M9W+GQ4vyDFEMctTstyG6GTcmZEBi8xcMnnXIEZUQ12Wf7QgbWnSUgkI6VlJ7A3aQMmCgmvcQg8xvCwDFfZEykHZchEtDhrTkMvsamEhRuhyW6EtVgfIa0uzY0BhBw/vanm77gXqtKqbCgzemNfDotNAQxBAQlFhNl7d99S1EH5VDyPKl5L0NfQaDBEXT44jwFwXVr/iPk3lERp6DB6i6VVPxgpf8Eup+ReazDSDTF6lKKHqjLeAhV4zBv8jmZfmVmTV9iGx60oYh--Lwdl39NO+FT9fIky--jwEXecDaCxbMtFoINC9p+Q==
```
Comment (styliii) [inline config/credentials.yml.enc:1]: Need a `secret_key_base` now

**Comment by styliii** [comment]
@keithtom i kept a lot of extra devise related code in here, in case you think we need some of these other options. This is just bare bones authentication, but helps derisk things a bit.

File: app/controllers/users/registrations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+  # GET /resource/sign_up
+  # def new
+  #   super
+  # end
+
+  # bug in devise for rails 7 https://github.com/heartcombo/devise/issues/5443
+  def sign_up(resource_name, resource)
+    sign_in(resource_name, resource, store: false)
+  end
```
Comment (keithtom) [inline app/controllers/users/registrations_controller.rb:16]: Just FYI: eventually we'll let people sign up but for now they are invite only, sorry should have mentioned this.

File: app/models/user.rb [hunk 3]
```diff
@@ -2,6 +2,11 @@
 
 class User < ApplicationRecord
   include ApplicationRecord::ExternalIdentifier
+  include Devise::JWT::RevocationStrategies::JTIMatcher
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+  :validatable, :jwt_authenticatable, jwt_revocation_strategy: self
```
Comment (keithtom) [inline app/models/user.rb:9]: cool.  I didn't know this existed.

File: app/controllers/users/registrations_controller.rb (hunk 2 above)
Comment (keithtom) [reply app/controllers/users/registrations_controller.rb:16]: We can jsut keep signup but keep it hidden for now too

**Review by keithtom — APPROVED** [review]
Seems good to me!  I haven't looked at the devise JWT but assumes it does the thing we need it to!   Is the auth stored in a cookie?  I thought there was a setting needed to make it work with multiple domains like platform.wildflowerschools.org and api.wildflowerschools.org

**Comment by styliii** [comment]
> Seems good to me! I haven't looked at the devise JWT but assumes it does the thing we need it to! Is the auth stored in a cookie? I thought there was a setting needed to make it work with multiple domains like platform.wildflowerschools.org and api.wildflowerschools.org

No, since it's using a JSON web token, instead of a cookie, we side step the need to make it work with multiple domains. At sign up, a JWT is created in the authorization header. The front end should then store the token in a HTTP only cookie. Any future requests made should include the token in the Authorization header. I'm pretty sure we don't need to do anything special to make it work across domains, but i could be wrong.

File: app/controllers/users/registrations_controller.rb (hunk 2 above)
Comment (styliii) [reply app/controllers/users/registrations_controller.rb:16]: ok, so we they'll receive an email for sign up? then their password will be created? so i'm guessing the user record is created before the password is?

Comment (keithtom) [reply app/controllers/users/registrations_controller.rb:16]: We'll create all the user objects with a spreadsheet.   i think without password.  they get invited via email with a special token that authenticates them automatically (without password).   that's just token auth, there's a devise plugin but also super easy to just code up.    and we can support normal email/password auth of course too.

**Pull request merged**